python/ray/experimental/iter.py

```diff
@@ -0,0 +1,556 @@
+
+    def num_shards(self) -> int:
+        """Return the number of worker actors backing this iterator."""
+        return sum(len(a.actors) for a in self.actor_sets)
+
+    def get_shard(self, shard_index: int) -> "LocalIterator[T]":
+        """Return a local iterator for the given shard.
+
+        The iterator is guaranteed to be serializable and can be passed to
+        remote tasks or actors.
+        """
+        a, t = None, None
+        i = shard_index
+        for actor_set in self.actor_sets:
+            if i < len(actor_set.actors):
+                a = actor_set.actors[i]
+                t = actor_set.transforms
+                break
+            else:
+                i -= len(actor_set.actors)
+        if a is None:
+            raise ValueError("Shard index out of range", shard_index,
+                             self.num_shards())
+        return _SingleActorIterator(a, t)
+
+
+class LocalIterator(Generic[T]):
+    """An iterator over a single shard of data.
+
+    It implements similar transformations as ParIterator[T], but the transforms
+    will be applied locally and not remotely in parallel.
+
+    This class is **serializable** and can be passed to other remote
+    tasks and actors. However, it should be read from at most one process at
+    a time."""
+
+    def __init__(self,
+                 base_iterator: Callable[[], Iterable[T]],
+                 local_transforms: List[Callable[[Iterable], Any]] = None,
+                 timeout: int = None,
+                 name=None):
+        self.base_iterator = base_iterator
+        self.built_iterator = None
+        self.local_transforms = local_transforms or []
+        self.timeout = timeout
+        if name is None:
+            name = "LocalIterator[?]"
+        self.name = name
+
+    def _build_once(self):
+        if self.built_iterator is None:
+            it = iter(self.base_iterator(self.timeout))
+            for fn in self.local_transforms:
+                it = fn(it)
+            self.built_iterator = it
+
+    def __iter__(self):
+        self._build_once()
+        return self.built_iterator
+
+    def __next__(self):
+        self._build_once()
+        return next(self.built_iterator)
+
+    def __str__(self):
+        return self.name
+
+    def __repr__(self):
+        return self.name
+
+    def for_each(self, fn: Callable[[T], T]) -> "LocalIterator[T]":
+        def apply_foreach(it):
+            for item in it:
+                if isinstance(item, YieldIterator):
+                    yield item
+                else:
+                    yield fn(item)
+
+        return LocalIterator(
+            self.base_iterator,
+            self.local_transforms + [apply_foreach],
+            name=self.name + ".for_each()")
+
+    def filter(self, fn: Callable[[T], bool]) -> "LocalIterator[T]":
+        def apply_filter(it):
+            for item in it:
+                if isinstance(item, YieldIterator):
```

Comment:
```suggestion
                if isinstance(item, YieldIterator) or fn(item):
```

Reply:
What does the `YieldIterator` case mean here? Are we just propagating it up? In general I find the `YieldIterator` semantics a little bit confusing as it seems to be used differently in difference places. Would be good to add some comments around that.

Reply:
I renamed it to NextValueNotReady, which should make it more clear it's a kind of timeout error.